zaza/openstack/charm_tests/saml_mellon/tests.py

```diff
@@ -156,3 +158,177 @@ def _do_redirect_check(url, region, idp_expect, horizon_expect):
+        else:
+            unit = zaza.model.get_units("openstack-dashboard")[0]
+            horizon_ip = unit.public_address
+
+        if self.tls_rid:
+            proto = "https"
+        else:
+            proto = "http"
+
+        # Use Keystone URL for < Focal
+        if self.current_release < self.FOCAL_USSURI:
+            region = "{}://{}:5000/v3".format(proto, keystone_ip)
+        else:
+            region = "default"
+
+        idp_address = zaza.model.get_units(
+            self.test_saml_idp_app_name).pop().data['public-address']
+
+        url = "{}://{}/horizon/auth/login/".format(proto, horizon_ip)
+        horizon_expect = '<option value="{0}">{1}</option>'.format(
+            self.horizon_idp_option_name, self.horizon_idp_display_name)
+
+        # This is the message the local test-saml-idp displays after you are
+        # redirected. It shows we have been directed to:
+        # horizon -> keystone -> test-saml-idp
+        idp_expect = (
+            "A service has requested you to authenticate yourself. Please "
+            "enter your username and password in the form below.")
+
+        def _do_redirect_check(url, region, idp_expect,
+                               horizon_expect, idp_address):
+
+            # start session, get csrftoken
+            client = requests.session()
+            # Verify=False see note below
+            login_page = client.get(url, verify=False)
+
+            # Validate SAML method is available
+            assert horizon_expect in login_page.text
+
+            # Get cookie
+            if "csrftoken" in client.cookies:
+                csrftoken = client.cookies["csrftoken"]
+            else:
+                raise Exception("Missing csrftoken")
+
+            # Build and send post request
+            form_data = {
+                "auth_type": self.horizon_idp_option_name,
+                "csrfmiddlewaretoken": csrftoken,
+                "next": "/horizon/project/api_access",
+                "region": region,
+            }
+
+            # Verify=False due to CA certificate bundles.
+            # If we don't set it validation fails for keystone/horizon
+            # We would have to install the keystone CA onto the system
+            # to validate end to end.
+            response = client.post(
+                url, data=form_data,
+                headers={"Referer": url},
+                allow_redirects=True,
+                verify=False)
+
+            if idp_expect not in response.text:
+                msg = "FAILURE code={} text={}".format(response, response.text)
+                # Raise a custom exception.
+                raise FailedToReachIDP(msg)
+
+            idp_url = ("http://{0}/simplesaml/"
+                       "module.php/core/loginuserpass.php").format(idp_address)
+
+            # Validate that we were redirected to the proper IdP
+            assert response.url.startswith(idp_url)
+            assert idp_url in response.text
```

Comment:
This is a big function to be embedded in another.  Please break it out and add `@staticmethod` to it, plus docstring with purpose, parameters and types.

Reply:
This function was carried on from the previous implementation.

I agree with you. I'll clean it up.